### DISCUSSION TITLE
Choice of Go vs Rust

### DISCUSSION THREAD

**Discussion by pkazmier** [zk-org/zk#248] [General]
Hi Mickael,

Somewhat off-topic, but I could find no others means of contacting you. I was somewhat curious as to the reason you chose Go for `zk` instead of Rust. I believe you had mentioned this was your first Go project. And yesterday, while searching for some help on serde (Rust serialization library for those unaware), I saw your name pop up with an answer. So, I’m curious, given you are proficient in Rust as well, what motivated you to use Go back then as Rust would have been a good choice as well for a CLI tool. And now a couple of years later, do you wish you had made an alternate choice?

I’m trying to learn Rust now, having tried Go, which I did not like due to its lack of expressiveness and verbosity as one that leans heavily to functional programming. I love Go’s concurrency model though where they hide async from the programmer and allow one to use a synchronous programming model. I’m not a fan of Rust’s async/await model (or any language that exposes that), and the new Rust Generic Keyword discussion the other day just reminds me of this further. Rust is the language I want to love, but just disappointed in the concurrency model. Go has the “right” model, but the language itself is just so blah.

I’d love to hear your thoughts on how you chose Go vs Rust for `zk` if/when you have time.

**Comment by mickael-menu** [comment]
My background is in mobile development, and I was not proficient in Rust nor Go when starting `zk`.

I initially chose Rust for this project and had a lot of fun as it's a really unique language with interesting concepts. However, after a month messing with it I grew frustrated by the productivity hit induced by learning Rust. I had a lot of unlearning to do, as many patterns common in other languages are just impossible in Rust.

Given I was learning Rust to create `zk` - which I needed - and not the other way around, I decided to pivot to Go. It has a lot of warts as you mentioned, but it was a breeze and the project went probably 10x faster.

Here's the state of the Rust version of `zk` when I stopped, if you're interested: [zk-rs-main.zip](https://github.com/mickael-menu/zk/files/9217168/zk-rs-main.zip)

As I lean more towards functional programming too, if I would start from scratch *now* I would probably use [Kotlin/Native](https://kotlinlang.org/docs/native-overview.html). It was actually my first choice but at the time it was not easily cross-platform.
